Add design document

mark-burnett opened the pull request:
This is the initial design document for Deckhand.  It specifies the major features and API endpoints in some detail and includes some examples.

mark-burnett committed:
WIP: Add initial design document

sh8121att commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+This type of metadata allows the following metadata hierarchy:
+
+* `name` - string, required - Unique within a revision for a given `schema`.
+* `storagePolicy` - string, required - Either `cleartext` or `encrypted`.
```
Maybe note that only the data section will be encrypted, the metadata section will be persisted in cleartext so Deckhand can include the document in merge/substitution operations.

mark-burnett replied:
Clarified.

sh8121att commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+* `substitutions` - list, optional - A sequence of substitutions to apply. See
+  the Substitutions section for additional details.
+  * `dest` - dict, required - A description of the inserted content destination.
+    * `path` - string, required - The JSON path where the data will be placed
```
I don't recall - did we conclude that a `pattern` field was unnecessary and intra-field substitution was not permitted?

mark-burnett replied:
I accidentally omitted it, thanks for the catch!

sh8121att commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+    - site-type
+    - region
+    - site
+    - force
```
Pedantic, but maybe include the explicit end document tag `...` on the examples to show the document doesn't continue.

mark-burnett replied:
Done.

bryan-strassner commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+
+This document defines the strict order in which documents are merged together
+from their component parts. It should result in a validation error if a
+document refers to a layer not specified in the `LayeringPolicy`.
```
Thoughts:
1) Documents should be able to be loaded that reference a non-existent layer - yes, they should fail at some stage of validation - probably initial validation, but still be ok to load and store.
2) If a revision of the layering policy is loaded that eliminates a layer, this could trigger the need for a LOT of tombstone document loads so as to be able to pass validation. Perhaps there's a future optimization that would auto-tombstone?
3) I'm thinking the answer is no to this: should a validation error occur if the layering policy has layers that NO documents reference?

sh8121att replied:
IMO
1. Agreed - cares for the ordering issue
2. An alternative could be only allowing layeringPolicy updates to re-order or add layers.
3. Also think this should be no as I can see a very small set of layering policies in the entire ecosystem and some sites may use minimal layering.

mark-burnett replied:
1.  Agree, was not planning to reject requests on such validations.  Maybe we could have a flag that would control rejection?  Or even allow non-committing dry-run posts that execute validations right away?
2. Could be useful, but I'd like to punt for now if we can until we understand usage patterns better.
3. It seems fine to have unused layers in some sites.  I'm imagining having quite a few layers being nice for flexibility, which would increase the likelihood of unused layers.

bryan-strassner replied:
1. I think that it is reasonable to issue warnings on the POST, and let the client ignore them all they want. 
200 OK
{ ignorable_warnings: [ ... ] }
(or whatever)

The service should diligently store the document as given. 
The stricter validations later will take care of things if they don't get settled by subsequent POSTs.

Dry run posts seem like feature creep, and would beg the question of wanting to have a whole series of POSTs run together as a dry run - definitely doable, but probably not high priority right now.  If we were going to do this, I think I'd rather allow for deletions of "unverified" versions,  maybe tied to some final validation that locks a version as permanent.

bryan-strassner commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+
+This type of metadata allows the following metadata hierarchy:
+
+* `name` - string, required - Unique within a revision for a given `schema`.
```
are there constraints on name minimum & maximum length?  Same kind of question on several subsequent fields here.

mark-burnett replied:
Good question.  I'm OK with leaving that out of this document for now and letting the JSON schema author (probably @fmontei) decide at implementation time.

bryan-strassner commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+    layer: region
+    parentSelector:
+      labels: parents
+      must: have
```
?  'must' ? I think formatting is messed here.

mark-burnett replied:
Actually, "must" was mean to be sample data.  The idea is that `parentSelector` is a collection of key + value pairs that must all be present in the parents for them to be selected.

I'll chose different sample data to avoid confusion.  We'll also have to clarify the selector/label paradigm a bit more.

bryan-strassner commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+* `replace` - overwrite data at the specified path and replace it with the data
+  given in this document
+* `delete` - remove the data at the specified path
+
```
If a document contains data that has no paths that have actions associated with them in the layering metadata, is there a default behavior of merge or is this a validation error? At the surface, I think we'd benefit from the default of merge, but still allow for merge commands to manage some oddities of the application of the actions.

If I have multiple actions that reference the same or nested paths, is this a validation error, or do we have complex interactions? ( the ordering might make this manageable, but I suspect it's not quite trivial)

sh8121att replied:
I feel like layerPolicy-based precedence should apply to data and merge-actions. And the rules of only allowing multiple source documents in a layer if there is no conflict should also apply to merge-actions on overlapping data paths.

mark-burnett replied:
Just restating all the questions and adding some related ones here for my own clarity:

1.  Is it an error to have additional data in a merge document that is not specified in at least one action?
2. Should we have a default behavior when no actions are specified?
3. If we have a default action, should we apply it even when there are other actions?
4. Do we allow action paths to overlap?
5. Do we allow multiple merge target parents from the same layer?

My thoughts on each:

1. I was initially thinking to ignore it, but maybe it should be an error?  I'm not 100% clear if this is a tricky one to check or not.
2. I don't think it's needed, assuming we allow paths to overlap (question 4), since in that case you can just add a `method=merge, path=.` action.
3. I'm torn on this, again relates to action path overlap.
4. I think there's not much harm in allowing this since it's fully deterministic and still has to pass schema validation after it occurs.
5. I would like to avoid allowing multiple targets from the same layer, since the main push for that seemed to be related to Armada using a single large document, and that's going to change soon.

I'm feeling a bit less confident about the merge chain construction at this point.  I would really like to avoid a proper DAG or even a tree and just have a clear, linear path (maybe specifying just parent name or similar).  I guess we could implement that by limiting the parent selection to the highest precedence (latest order) document that matches the selector (that is also in a layer above the current document).  If we can limit it to a single parent, I think it becomes simpler to manage.  Thoughts?

bryan-strassner replied:
for (1) I think it's going to be unintuitive if layered documents don't result in composition as a default (i.e. merge everything). I'm also willing to accept that I am misunderstanding what's going on here. 
~~However, the way I'd see this working if we default merged, all the data in the document is meaningful, and there's less confusion than if things end up ignored.  This means that the two primary actions left are replace and delete, which can be applied, and overrule any prior merging (I think this is the natural and expected outcome of these actions - no magic here). What I still haven't wrapped my head around is if we'd still need a merge command at all - since I think a merge after a replace or delete can both be seemingly be done with a single, more explicit replace in the first place.  (So, I'm thinking we wouldn't need or want a merge command)~~

... upon further thought. If at each layer during the merge we were to apply all deletes before doing a wholesale merge of the data, we'd have a complete representation we could work with, would still be deterministic, and perhaps understandable to someone lookng at it. I'm pretty sure we've discussed this in person, but replace is easily indicated by delete + merge. If all deletes happened first against the parent document, then merges occur, there are several advantages:
1)  The actions in the metadata is simply a list of nodes to delete. (recommend naming it "deletes" or some such)
2) The order of the deletions may occur in any order (as long as they are all before the merge step.
3) the Data section is all important since it is not reliant on actions that accompany it.
4) there is no worry about overlapping paths (delete doesn't care, and a well formed document prevents it)

mark-burnett replied:
After some more conversation, the direction we are looking at is to only allow a single parent at each layer, which avoids some of the concerns here.

sh8121att replied:
I think enforcing a single parent is a good start for 1710 and if we find use case gaps, we have some alternative designs that can be used.

sh8121att commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+  name: site-deploy-ready
+data:
+  validations:
+    - name: deckhand-schema-validation
```
Do we want some description of the validation so Deckhand knows how to execute a revision - maybe a URL that Deckhand would POST a revision reference and enough data that the receiving service knows how to post back the validation result? Or is this Shipyard's domain?

mark-burnett replied:
It feels more like Shipyard's responsibility, but I could be convinced we should add some kind of hook registration system.

sh8121att commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+
+## Validation
+
+Services can report success for validation types for a given revision.
```
A better place for my above question for how validations get initiated.

sh8121att reviewed (commented):
Great work Mark on bringing lots of tangential discussions into a coherent document. I can work on some of the figures that are noted in TODOs.

mark-burnett commented:
Thanks for the early feedback @bryan-strassner and @sh8121att.  I will make sure to address all comments this weekend before expanding on additional points.

mark-burnett committed:
Address some initial review comments

mark-burnett committed:
Add samples for remaining endpoints

mark-burnett committed:
Move reporting of validation status

Moving status indications of individual validations into the tree of
each validationPolicy is necessary, because the validation may have
different expiration times for the different policies that rely on it.

mark-burnett commented on docs/design.md:
```diff
@@ -0,0 +1,547 @@
+  select `promenade/Node/v1` documents, and `schema=prom` would not select
+  `promenade` documents.
+* `metadata.name` - string, optional
+* `metadata.layeringDefinition.abstract` - string, optional - Valid values are
```
@bryan-strassner I don't think I like this.  It seems like we might need separate endpoints for rendering full documents and for getting the raw, unmerged, unsubstituted documents for administrative purposes.

bryan-strassner replied:
since we are POSTing documents, it seems like GET documents should return those things being posted (Doesn't have to, but it is user friendly to do so)- so I agree, a new endpoint for compiledDocuments seems like a very reasonable thing to GET.  I also think it's reasonable that a compileDocument include references somehow to the list of all documents that went into that compilation.

sh8121att replied:
+1 to @bryan-strassner's take

mark-burnett replied:
Implemented -- see the `/revisions/{revision_id}/rendered-documents` endpoint.

mark-burnett committed:
Provide a separate rendered-documents endpoint

This gives a clearer separation between administrative document queries
and typical requests for fully rendered configuration.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+    control document.
+  * `parentSelector` - labels, optional - Used to construct document chains for
+    executing merges.  See the Layering section below for details.
+  * `actions` - list, optional - A sequence of actions to apply this documents
```
document's rather than documents

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+    control document.
+  * `parentSelector` - labels, optional - Used to construct document chains for
+    executing merges.  See the Layering section below for details.
+  * `actions` - list, optional - A sequence of actions to apply this documents
```
Should actually be "to apply to this document's data". You could also include: "See the Layering section for additional details" where you specify the list of actions.

mark-burnett replied:
I moved the "see Layering section" up to the `layeringDefinition` key to try and help clarify this.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+          ignored:  # Will not be part of the resultant document after layering.
+            data: here
+  substitution:
+    target: null  # Paths do not need to exist to be specified as substitutiond destinations.
```
Perhaps this comment makes more sense on L118 above?

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+metadata:
+  schema: metadata/Document/v1
+  name: application-api
+  storagePolicy: cleartext
```
Shouldn't this be encrypted?

mark-burnett replied:
Well, it's not a secret, so it wouldn't have to be, but if policy required it, it would be fine to do here.  This is configurable per document.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,431 @@
+documents.
+
+Revisions can be thought of as commits in a linear git history, thus looking
+at a revision includes all content from previous revision.
```
from previous revisions

mark-burnett replied:
Fixed, thanks.

mark-burnett committed:
fix typo

mark-burnett committed:
Add overall layering description

mark-burnett committed:
Reference Layering section in layeringDefinition description

mark-burnett committed:
Add description of substitution.

intlabs commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+* `schema` - Defines the name of the JSON schema to be used for validation.
+  Must have the form: `<namespace>/<kind>/<version>`, where the meaning of
+  each component is:
+  * `namespace` - Identifies the owner of this type of document. The
```
is `namespace` the most appropriate label to use here? the description would seem to make `owner` more appropriate?

mark-burnett replied:
Thanks for taking the time to look at this, @intlabs, I know you're plate is full.

I agree that `owner` is a tempting name and would be happy to switch.  The only catch with that particular name is that we use the same concept for both `schema` and `metadata.schema`, where currently the only supported namespace/owner is "metadata".

Maybe we should simply omit the prefix for `metadata.schema`, since it's unclear it would ever be needed?

Would anyone else care to weigh in?

sh8121att replied:
I think `namespace` is a clear definition of what the field is used for. In my mind `owner` has a connotation of authorization for management which this field is not driving. In the end I think it is not overly important as the document field will actually be `schema` and the format breakdown to constituent parts  is only used in documentation.

fmontei replied:
I agree that `namespace` is preferable to `owner`.

intlabs commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+
+* `name` - string, required - Unique within a revision for a given `schema`.
+* `storagePolicy` - string, required - Either `cleartext` or `encrypted`. If
+  `encyrpted` is specified, then the `data` section of the document will be
```
Will deckhand support different encryption types/containers? eg symmetric/certificate etc.

mark-burnett replied:
We talked about this a bit, and couldn't come up with a particular justification for not just using whatever the strictest encryption policy applicable to the site for all its data, so we didn't add hooks for controlling it.

I think we could make it more flexible in the future in a backwards-compatible way, e.g. by moving to something like `storagePolicy: encrypted+<option>`.

Did you have a specific use case in mind?

intlabs replied:
Mostly as this expected of most key-managers I've worked with - it is not about the 'strictness' per-say of the data store - but more about having useful metadata within it. I could, however, foresee cases where you may want to have some data stored via an asymmetric key and others via symmetric depending on where the data in question is generated, and hence loaded into the secret store - depending on the RBAC/ABAC and trust model intended. Would that access policy (and authorization) fall within the scope of this document? currently, it looks like the model would assume that all users are either granted access to all data stored within deckhands backing store or that it's assumed each (Keystone)project would operate in its own world?

mark-burnett replied:
It's actually the opposite.  Users of deckhand are expected to not have access to any of the data in the backing store (or even know what the backing store is).  Any necessary access controls will be handled by Deckhand's authorization policies.

It's possible that in the future we will choose to add per-document access policy, but it seems to be a speculative idea with no immediate, concrete use case that I'm aware of.  Do you know of a concrete use case for this feature?

intlabs replied:
What I'm referring to is that deckhand authorization policy. A per-document may be overly restrictive - but as it stands it would appear that a service (or user) would be able to request credentials and keys that have nothing to do with it - as deckhand - though storing secrets securely, would appear to hand them out to anyone who requested them.

sh8121att replied:
I believe the current perspective is that Deckhand will enforce authorization at the action level making all documents equivalent. 'Anyone who requested them' is a bit broad as I believe we plan to have service accounts for all the UCP services accessing deckhand, so RBAC will enforce that those service accounts do or do not have access to the data ecosystem. But if a service account does have authority for an action, it can execute that action against any document (including secrets). Otherwise Deckhand's policy needs to have intrinsic knowledge of what downstream services might be doing.

alanmeadows replied:
I think a per document specification makes a lot of sense but I really want to see what this looks like. Are we specifying roles? How is read and write of this particular kind for this particular role specified? What will the user experience be (auth error, empty returns (hidden elements), etc.)  What will the merge experience be with conflicting documentation authorization specs.  On the positive side, this approach means there is no implied security - its very explicit spelled out in the document.

To be sure, I think this approach can align with what I imagine is the goal here with deckhand; the more we create very UCP-centric views the more it has to constantly adapt to new UCP elements/models in its API. If we instead are treating it as a generic repository it doesn't need to have a sprawling list of specific views with RBAC associated with them as the documents self-manage that.  That said, there is still no getting around spelling this out somehow -- that same UCP-centric specification just ends up requiring a thicker policy.yaml for deckhand with roles for each service articulated.

This also implies in some way that service-level authentication is occurring for the security to actually have any value -- initially we said that wasn't likely going to fit within the time frame and settled on Shipyard passing down its "long-lived" token throughout the chain.  Service accounts are of course more clean but we need to make sure we can implement it in a reasonable time frame.  In the end however, it doesn't matter either way to the deckhand framework itself - either we have a single shipyard role with very open permissions or we have achieved service level authentication where each service acts on behalf of itself and we can start closing things up into an as-needed-basis.

sh8121att replied:
I think we did settle on needing service accounts because otherwise the services cannot enforce RBAC with a token. Keystone requires an authenticated session to yield roles for a token, so each UCP service needs at least enough authority to get account name and roles for a client-provided token.

sh8121att replied:
An idea for how this per-document authorization could be implemented.
This assumes each UCP service (and anyone else using Deckhand) has a UCP-Keystone service account with assigned roles similar to what I've sketched out

* promenade: 'physical_provisioner', 'host_configurator'
* drydock: 'physical_provisioner'
* shipyard: 'orchestrator', 'data_ingress'
* armada: 'container_provisioner'

The document access policy would be respected through the entire merge/substitution process. If a source document was not allowed, metadata would be retained in the result document. If omitting the source from the compiled result document still allowed the result document to pass schema validation, another Deckhand validation policy could determine if running into a disallowed source during merge/substitution was considered invalid. If the result document couldn't pass schema validation with the source document omitted, this would obviously fail validation

```
schema: deckhand/AccessPolicy/v1
metadata:
  schema: metadata/Document/v1
  name: ucp_open
  labels:
    accesspolicy: default
  layeringDefinition:
    layer: global
data:
    document_read:
      allowed_roles:
        - 'physical_provisioner' #Drydock, Promenade
        - 'orchestrator' #Shipyard
        - 'container_provisioner' #Armada
    document_write:
      allowed_roles:
        - 'data_ingress' #Shipyard
    document_admin:
      allowed_roles:
        - 'data_ingress' # Shipyard
        - 'admin' # Individual account for cleanup and debugging
```

```
schema: deckhand/AccessPolicy/v1
metadata:
  schema: metadata/Document/v1
  name: armada_sensitive
  labels:
    accesspolicy: restricted
    accesstarget: armada
  layeringDefinition:
    parentSelector:
      accesspolicy: default
    actions:
      - method: replace
         path: .document_read
data:
  document_read:
    allowed_roles:
      - 'container_orchestrator'
```
```
schema: deckhand/CertificateKey/v1
metadata:
  schema: metadata/Document/v1
  name: keystone_api_tls_key
  accessPolicy: armada_sensitive
data: |
  SOME_KEY_DATA
```

intlabs replied:
@sh8121att that schema is looking pretty nice. With some clarification as to what `physical_provisioner`et all refer to (role/project/user - though I'm assuming role atm) I think this, or another similar, approach alleviates my concerns.

mark-burnett replied:
I think either a centralized access policy approach or an individual
document-level approach could be fine.

Here's another example of access policy, with a slightly different format (and
slightly less flexible as it does not allow for layering or substitution):

```yaml
---
schema: deckhand/DocumentAccessPolicy/v1
metadata:
  schema: metadata/Control/v1
  name: document-access-policy
data:
  rules:
    - name: descriptive-name-01
      selector:
        foo: bar
      readRoles:
        - drydock
        - promenade
---
schema: deckhand/ValidationAccessPolicy/v1
metadata:
  schema: metadata/Control/v1
  name: validation-access-policy
data:
  armada-validation-01:
    writeRoles:
      - armada
---
schema: deckhand/ValidationPolicy/v1
metadata:
  schema: metadata/Control/v1
  name: validation-policy
data:
  readRoles:
    - shipyard
  validations:
    - name: same-format-as-before
...
```

On document read, you would iterate over the list of rules until finding a
successful match.  Any match would allow access.  Falling off the end of the
list with no hits would deny access.

`ValidationAccessPolicy` indicates which roles are allowed to post validation
updates for given revisions, and the new field in `ValidationPolicy` specifies
what roles are allowed to check the status of a validation policy and
transitively all its validations (this avoids keeping the two authorization
definitions consistent).

Here's an example of per-document policies for comparison:

```yaml
---
schema: deckhand/CertificateKey/v1
metadata:
  schema: metadata/Document/v1
  name: some-key
  storagePolicy: encrypted
  accessPolicy:
    readRoles:
      - armada
data: |
  SOME PRIVATE KEY
---
schema: deckhand/Validation/v1
metadata:
  schema: metadata/Control/v1
  name: armada-validation-01
data:
  writeRoles:
    - armada
---
schema: deckhand/ValidationPolicy/v1
metadata:
  schema: metadata/Control/v1
  name: validation-policy
data:
  readRoles:
    - shipyard
  validations:
    - name: same-format-as-before
...
```

In either case, I think we're only interested in specifying read access via
this metadata.  Write access is more complicated, because inserting a document
can have side effects due to the layering system; therefore, I think we should
configure write access to be limited to roles defined directly in deckhand
configuration.

I would think we would want to start with a centralized access policy and
consider adding per-document policy to override these if we find they are
needed.

Either of these approaches could be added in a backwards-compatible way.

fmontei replied:
Regarding: "either a centralized access policy approach or an individual document-level approach could be fine".

Centralized access policy is what all of OpenStack does and is much easier to debug and implement, particularly because ``oslo.policy`` can be leveraged to accomplish that. The advantages are pretty straightforward: 

- easier to integrate/implement than per-document RBAC
- self-documenting
- tried and true implementation of RBAC

Per-document RBAC would require retrieving the access policy from the DB and parsing it to figure out whether a particular read, delete or write operation is allowed. For example, it seems inefficient for a role to have read and write permissions but not delete permissions, so this means we could get 50% of the way through an intricate YAML merge operation only to fail on some delete operation. It might be possible to figure out in advance all the operations needed to completely merge a YAML document, but clearly per-document RBAC is a much more difficult thing to implement and debug. That is, if there is an issue with the access policy itself and not the YAML file, then the error thrown by Deckhand would probably have to return the access policy information in addition to where it failed in the YAML merge procedure just in order for a user to figure out the source of the problem.

On the other hand, per-document RBAC is dynamic and far stricter than a more centralized RBAC approach using ``oslo.policy``.

sh8121att replied:
@fmontei I think we'll end up with a oslo.policy-like RBAC system for 1710. But do we need to upgrade to a per-document model in the future to provide features desired by CSO or the community?

mark-burnett replied:
@fmontei Thanks for your input.  I am perfectly happy starting with a centralized policy and just using a wait-and-see if we need to offer more flexibility.

@alanmeadows Are you satisfied with a centralized policy-based design resembling the above?  If so, I'll write it up a bit more formally and add some details about validation/errors.

alanmeadows replied:
I appreciate all the comments on the complexities.  I have no issue with an "oslo.policy-like" RBAC implementation for now.  We may even find we don't need to pivot from this even post-1710.

I think we need some concrete examples on how the `selector:` pattern would work.  There were a few ideas/styles in everyones examples that I think need to get merged into a single view. 

That said, since we have no actual need of a dynamic policy API to allow us to adjust on the fly for the undercloud given we manage policies very predictably with helm, will we simply leave `enable_centralized_policy` disabled in oslo.policy? I want us to be cognizant of not requiring community end users to deploy a spaghetti web of dependent services in order to use any single standalone project on its own.

alanmeadows replied:
@sh8121att Makes sense to me on the service account front. Will be a lot easier to explain: It works like OpenStack. Move along.

alanmeadows replied:
We likely need to plan the work necessary to register the endpoints of various services (and lock in on their service category names (openstack equivalents being glance: image; nova: compute, so on) into keystone so that both clients and potentially cross-service communication can leverage the IAM to resolve other services endpoints.  The only configuration they will need is the Underclouds Keystone details.  This does make Keystone a dependency of every undercloud service but I would like to think everyone wants authentication and RBAC.

sh8121att reviewed (approved):
LGTM

alanmeadows commented:
@sh8121att Makes sense to me on the service account front.  Will be a lot easier to explain: It works like OpenStack. Move along.

fmontei commented on docs/design.md:
````diff
@@ -0,0 +1,785 @@
+...
+```
+
+If `region-1234` were later removed, then the parent chosen for `site-1234`
````
Thanks for the detailed examples, this is really helpful.

mark-burnett replied:
Welcome!  We will probably want to add more as we discover more corner cases or surprising behavior.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+Selection of document parents is controlled by the `parentSelector` field and
+works as follows. A given document, `C`, that specifies a `parentSelector`
+will have exactly one parent, `P`. Document `P` will be the highest
+precedence (i.e. part of the lowest layer) document that has the labels
```
I had to read this a few times. The order of precedence is defined by data.layerOrder in the LayeringPolicy, so now this makes sense.

mark-burnett replied:
Quite right, I will try to update the language to be clearer.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+
+Substitution is primarily designed as a mechanism for inserting secrets into
+configuration documents, but works for unencrypted source documents as well.
+Substitution is applied at each layer after all merge actions occur.
```
If during the merge phase sections of the finalized document whose data section has been changed, then this could mean that previous references to sections in the data section by lower-precedence documents are no longer valid, i.e. point to non-existent sections. Is this something that is ignored (i.e. not treated as an error)?

mark-burnett replied:
It seems like the interplay between substitution and layering should be clarified more.  Let me take a stab at some thoughts that I could try to include in the doc:

Layering order is only tangentially related to substitution.  The key constraints on substitutions from my perspective are:

1. Any part defined of any fully-rendered, concrete (not abstract) document is a valid substitution source.  The layer of that document is not a direct part of the substitution process.
2. When rendering a document, at each layer:
  1. Layering actions are applied
  2. Substitutions are applied

This combination provides a couple of properties:

- Due to the constraints on layering (can only select parents from lower precedence layers, can only have one parent), that portion of the compute graph cannot have cycles.
- Due to the constraint that we can only source substitutions from fully rendered documents, we likewise cannot have cycles when including them -- this will probably need detection during validation.

And enables a couple of nice features:

- We can have site-specific things like secrets added to documents which are supposed to be uniform for a region (like Armada definitions).
- The final document generated is deterministic because the order of all operations is unambiguous (though non-trivial).

This description is based on a really a naive, recursive conceptual understanding.  Of course the implementation can be e.g. fully iterative or graph based -- not trying to imply such an implementation constraint.

Are these thoughts helpful?

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+  storagePolicy: cleartext
+  layeringDefinition:
+    layer: region
+  substitutions:
```
Is there a reason the `substitutions` section is left in the finalized document?

mark-burnett replied:
Great question.  @bryan-strassner and I had kicked around the idea of adding a new metadata schema for rendered documents that omitted this.  It might make sense for it to include additional details about what specific documents were involved in the layering and substitution process too. 
 What do you think about that general idea?

It's probably useful to return it on the non-rendered endpoint as it's part of the behavior definition.

fmontei replied:
I think that including information in the rendered document about how the document was rendered is a good idea. However, maybe it would be optional by passing a `?debug=True` query string or something? I'm not opposed to always returning this kind of information, it's just not immediately needed, except -- I would imagine -- for debugging purposes.

mark-burnett replied:
Yeah, having that behind a flag seems like a good idea.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+into other documents. This substitution is layer-independent, so given the 3
+layer example above, which includes `global`, `region` and `site` layers, a
+document in the `region` layer could insert data from a document in the
+`site` layer.
```
If a document with:

```yaml

layeringDefinition:
    layer: site
```
References a schema with `layer: region` using `substitutions[0].src` is this an error?

That is, if the order-of-precedence is violated, is this an error?

mark-burnett replied:
I think this is related to the above question about interplay between substitution and layering.  Probably need to add some clarifying comments to the document about those things.

In answer to your direct question no, this is not an error, it's actually a nice feature specifically for overriding things like secrets at a site level and still using a common configuration for things like armada chart definitions, etc.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+  storagePolicy: cleartext
+  layeringDefinition:
+    layer: region
+  substitutions:
```
If there are two substitutions that point to the same `dest.path` using different `src.schema` (that have the same `layeringDefinition.layer`) is that en error? Won't be clear which one to use for substitution.

mark-burnett replied:
Substitutions are applied in order and the layer of the source is ignored (except that it must be first fully rendered before it's used as a substitution source).

This means that there can in fact be conflict.  I think that since it's completely deterministic we might not want to forbid it right out the gate, thought I could quickly be talked into a deckhand-provided validation that checks for such scenarios.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+are special, and are used to control the behavior of Deckhand at runtime.  Only
+the following types of control documents are allowed.
+
+#### DataSchema
```
After reading this, still not sure what a `DataSchema` is. Is this meant to store different types of schemas referenced by Certificates, LayeringPolicies and so forth. Will the `DataSchema` define the validation schema (i.e. the jsonschema) needed to validate the format of the schema itself?

mark-burnett replied:
Good question!

So, ` DataSchema` is literally the jsonschema schema used to validate documents of the type its `metadata.name` matches.

Re: how to validate them, it will be the schema defined by `deckhand/DataSchema/v1`, which will essentially by the spec-standard jsonschema meta schema.

* [Here](http://json-schema.org/documentation.html) is the doc page for json schema
* [Here](http://json-schema.org/schema) is the direct link for the meta-schema

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+This endpoint is the only way to add, update, and delete documents. This
+triggers Deckhand's internal schema validations for all documents.
+
+If no changes are detected, a new revision should not be created. This allows
```
Are revisions created only for "rendered-documents"?

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+    url: https://deckhand/revisions/0
+    createdAt: 2017-07-14T21:23Z
+    validationPolicies:
+      site-deploy-validation:
```
Is `site-deploy-validation` an example of a particular validation or is it the overall status of the entire set of validations? If not, why not include just the overall status of the set of validations for each revision?

fmontei replied:
Line 666 below seems to suggest that this is the overall status of all the validations.

mark-burnett replied:
Sorry, it's meant to be an example.  I'm imagining defining a lot of specific validation policies that can quickly be queried to make sure that a particular action can be performed on a particular revision of the document set.

In this way, Shipyard could have a standard interaction pattern when it receives an action request:

1. Auth
2. Check Deckhand for the most recent revision that validates for the action
3. Execute the actual action using the chosen revision

Of course, such automatic revision detection isn't necessarily what will always be wanted.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+Validation status is always for the most recent entry for a given validation.
+A status of `missing` indicates that no entries have been created. A status
+of `expired` indicates that the validation had succeeded, but the
+`expiresAfter` limit specified in the `ValidationPolicy` has been exceeded.
```
Not clear how the status in:

```yaml

site-deploy-validation:
     url: https://deckhand/revisions/0/documents?schema=deckhand/ValidationPolicy/v1&name=site-deploy-validation
     status: failed
```
is calculated, assuming it's the overall status of the set of validations that were run. I'd imagine just 1 failed validation results in the overall failed validation status? What about "expired" or "missing"? Should there be any other statuses apart from "success" and "failure", i.e. "partial-success" or something?

mark-burnett replied:
Good question.  So individual validations can be missing, expired, failed or succeeded, but the overall validation policy can only be succeeded if all its comprising validations are succeeded.

This will give a binary go/no go response for applications that rely on these to define whether the configuration is in place for them to perform a particular action.

I'll update the document with a better description of the intent here.

fmontei commented on docs/design.md:
```diff
@@ -0,0 +1,785 @@
+---
+status: failed
+errors:
+  - documents:
```
The schema format below doesn't appear in the detailed description for a revision (under validations section), i.e. for "GET `/revisions/{{revision_id}}`" above. Shouldn't it?

mark-burnett committed:
Add clarifications based on review

mark-burnett committed:
Update API paths for consistency

mark-burnett committed:
Add basic RBAC details to design document

stannum-l commented on docs/design.md:
```diff
@@ -0,0 +1,865 @@
+There are 3 supported kinds of document metadata. Documents with `Document`
+metadata are the most common, and are used for normal configuration data.
+Documents with `Control` metadata are used to customize the behavior of
+Deckhand. Documents with `Tombstone` metadata are used to delete pre-existing
```
Is there a reason for a Tombstone document to delete than just do a DELETE /documents to remove the existing document?

mark-burnett committed:
Add DELETE endpoint

mark-burnett committed:
Clarify layering actions start from full parent data

sh8121att commented:
Is it time to get this merged as a baseline design so that ongoing tweaks can be managed by smaller, focused PRs?
